cryptoprotocol/src/main/java/de/upb/crypto/incentive/cryptoprotocol/model/Setup.java

```diff
@@ -0,0 +1,137 @@
+package de.upb.crypto.incentive.cryptoprotocol.model;
+
+
+import de.upb.crypto.craco.prf.PrfKey;
+import de.upb.crypto.craco.prf.aes.AesPseudorandomFunction;
+import de.upb.crypto.craco.sig.interfaces.SignatureKeyPair;
+import de.upb.crypto.craco.sig.sps.eq.SPSEQPublicParameters;
+import de.upb.crypto.craco.sig.sps.eq.SPSEQSignatureScheme;
+import de.upb.crypto.craco.sig.sps.eq.SPSEQSigningKey;
+import de.upb.crypto.craco.sig.sps.eq.SPSEQVerificationKey;
+import de.upb.crypto.incentive.cryptoprotocol.exceptions.PedersenException;
+import de.upb.crypto.incentive.cryptoprotocol.model.keys.provider.ProviderKeyPair;
+import de.upb.crypto.incentive.cryptoprotocol.model.keys.provider.ProviderPublicKey;
+import de.upb.crypto.incentive.cryptoprotocol.model.keys.provider.ProviderSecretKey;
+import de.upb.crypto.incentive.cryptoprotocol.model.keys.user.UserKeyPair;
+import de.upb.crypto.incentive.cryptoprotocol.model.keys.user.UserPublicKey;
+import de.upb.crypto.incentive.cryptoprotocol.model.keys.user.UserSecretKey;
+import de.upb.crypto.math.interfaces.hash.HashIntoStructure;
+import de.upb.crypto.math.interfaces.structures.GroupElement;
+import de.upb.crypto.math.pairings.generic.BilinearGroup;
+import de.upb.crypto.math.pairings.type3.bn.BarretoNaehrigBilinearGroup;
+import de.upb.crypto.math.structures.zn.Zn;
+import de.upb.crypto.math.structures.zn.Zn.ZnElement;
+
+/**
+ * a class implementing the incentive system algorithms (of the incentive system from the 2020 paper) that are not part of any protocol
+ * (namely Setup, P.KeyGen, U.KeyGen)
+ */
+public class Setup {
+    public static final int PRF_KEY_LENGTH = 256;
+
+    /**
+     * generates public parameters from security parameter
+     * @param securityParameter integer representation of security parameter
+     * @return public parameters as object representation
+     */
+    public static PublicParameters trustedSetup(int securityParameter) {
+        // generate a bilinear group from the security parameter (type 3, Barreto-Naehrig)
+        BilinearGroup bg = new BarretoNaehrigBilinearGroup(securityParameter);
+
+        HashIntoStructure hf = bg.getHashIntoG1(); // hash function {0,1}^* -> G_1
+
+        // compute w (base used in double spending protection, see 2020 incsys paper)
+        GroupElement w = (GroupElement) hf.hash("w" + bg.toString()); // TODO: ... .toString ok? in PR anmerken
```

Comment:
Why is this inconsistency a problem here? As far as I see, w and h7 are computed only once per lifetime of an incentive system instance.

Reply:
Indeed, in the current configuration, they're fixed once and never touched. 
More generally, the advantage of having the value be a hash of something would be that anyone can check whether the value has actually been computed using the hash (by just re-hashing the preimage). This ensures that, for example, the instance who generated the value isn't able to cheat and choose the value "cleverly" itself.
So here, we wouldn't necessarily fulfill this goal: if I spin up my own JVM, `bg.toString()` may output something completely different than the pp generator's JVM, so I really have no clue what the preimage shall be - hence I cannot check anything. 

So two choices: (1) take the stance that pp are generated honestly anyway, so no need to check anything (in this case, we can just as well have the trusted party choose the values randomly without hash), or (2) set up the scheme such that the issuer/store sets up these parameters (and may want to do so maliciously), in which case the hash is crucial, but then this whole generation should probably happen within the issuer key generation procedure instead of pp generation.

Reply:
Note: in the eprint paper version the generation of these values is somehow slipped into the Setup. This is not the intended definition, it was planned to be part of the issuer key generation. We will fix this in one of the ways described above in a future version of the paper.

Reply:
Thanks for the insight.
As far as I understand we should use hashing for having integrity guarantees about the public parameters, right?
The question is how to represent the BiliniearGroup as a string then. Should I do stuff like concatenating the integer representation of the generators of its groups or what do you suggest?

Reply:
Ideally, everything we hash is `UniqueByteRepresentable`, so one possible solution is to make `Group` implement `UniqueByteRepresentable`.
I guess we could do that.

On your end, right now, without changes to the `math` library, there is no great solution for hashing the `Group`. 

For now, I feel like the easiest fix is to keep the trusted pp generation and just remove hashing (have trusted party choose them randomly). Or what's your stance on this?

Reply:
That is definitely the easiest fix for sure.
So we let a trusted party randomly choose h7 and w right now and add some sophisticated hashing thing once the changes in the math library are done?